### PR TITLE
logging: uart: Add support for asynchronous API

### PR DESCRIPTION
**Pull request by nordic-krch**
Asynchronous API may utilize DMA and significantly reduce CPU load during outputing to UART.

A simple test was performed to estimate CPU load reduction. On `nrf52840dk` following loop was executed:
```
while (1) {
   LOG_INF("test");
   k_msleep(10);
}
```
CPU load was measured. For standard uart backend load was **42%** and for asynchronous API it was **0,5%**.

When `CONFIG_LOG_BACKEND_UART_ASYNC=y` but device does not support asynchronous API backend falls back to polling.

### CODE REVIEW

File: subsys/logging/log_backend_uart.c [hunk 1]
```diff
@@ -10,23 +10,38 @@
...
+	switch (evt->type) {
+	case UART_TX_DONE:
+		k_sem_give(&sem);
+		break;
+	default:
+		break;
+	}
+}
 
+static void dict_char_out(uint8_t *data, size_t length)
```
Comment (dcpleung) [inline subsys/logging/log_backend_uart.c:42]: Maybe rename this to `dict_char_out_hex()`? As this is only used for hexadecimal output.

File: subsys/logging/Kconfig.backends [hunk 2]
```diff
@@ -12,6 +12,10 @@ config LOG_BACKEND_UART
 
 if LOG_BACKEND_UART
 
+config LOG_BACKEND_UART_ASYNC
+	bool "Use UART Asynchronous API"
+	depends on UART_ASYNC_API
```
Comment (dcpleung) [inline subsys/logging/Kconfig.backends:17]: May also add `!CONFIG_LOG_BACKEND_UART_OUTPUT_DICTIONARY_HEX` as dictionary output in hex is still using polling for output.

File: subsys/logging/log_backend_uart.c (hunk 1 above)
Comment (nordic-krch) [reply subsys/logging/log_backend_uart.c]: done

File: subsys/logging/Kconfig.backends (hunk 2 above)
Comment (nordic-krch) [reply subsys/logging/Kconfig.backends:17]: done

**Review by dcpleung — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**